stage0/src/sev.rs

```diff
@@ -28,14 +28,64 @@ use oak_sev_guest::{
 use spinning_top::Spinlock;
 use x86_64::{
     instructions::tlb,
-    registers::control::Cr3,
+    registers::{control::Cr3, model_specific::Msr},
     structures::paging::{
         frame::PhysFrameRange, page::AddressNotAligned, Page, PageSize, PageTable, PageTableFlags,
         PhysFrame, Size1GiB, Size2MiB, Size4KiB,
     },
     PhysAddr, VirtAddr,
 };
 
+/// The identifier for the MSR used to set the default caching mode.
+const IA32_MTRR_DEFTYPE: u32 = 0x2ff;
+
+/// The cache memory type used with MTRR.  We only use Write-Protect mode which the Linux kernel
+/// expects to be enabled by the firmware in order to enable SEV.
+pub enum MemoryType {
+    WP = 5, // Write-Protect.
+}
+
+/// Models fields of the IA32_MTRR_DefType MSR.
+struct MTRRDefType {
+    mtrr_enable: bool,               // Set this to enable MTRR.
+    fixed_range_enable: bool,        // Set to enable fixed-range support.
+    default_memory_type: MemoryType, // The default caching mode.
+}
+
+impl From<MTRRDefType> for u64 {
+    fn from(deftype: MTRRDefType) -> Self {
+        let mtrr_enable = if deftype.mtrr_enable { 1 } else { 0 };
+        let mtrr_enable = mtrr_enable << 11; // MTRR enable bit is bit 11.
+        let fixed_range_enable = if deftype.fixed_range_enable { 1 } else { 0 };
+        let fixed_range_enable = fixed_range_enable << 10; // Fixed range nable is bit 10.
+        mtrr_enable | fixed_range_enable | deftype.default_memory_type as u64
+    }
+}
+
+/// Write a config value to IA32_MTRR_DefType MSR register to set the default caching mode to
+/// `default_type`.   The Linux kernel requires the mode be set to `MemoryType::WP` since July,
+/// 2022, with this requirement back-ported to 5.15.X, or it will silently crash when SEV is
+/// enabled.
+///
+/// The Linux kernel gives a warning that MTRR is not setup properly, which we can igore:
+/// [    0.120763] mtrr: your CPUs had inconsistent MTRRdefType settings
+/// [    0.121529] mtrr: probably your BIOS does not setup all CPUs.
+/// [    0.122245] mtrr: corrected configuration.
+pub fn enable(default_type: MemoryType) {
+    let enable_mtrr_in_wp_mode = MTRRDefType {
+        mtrr_enable: true,
+        fixed_range_enable: false,
+        default_memory_type: default_type,
+    };
+
+    // Safety: This operation is safe because this specific MSR has been supported since the P6
+    // family of Pentium processors (see https://en.wikipedia.org/wiki/Memory_type_range_register).
+    // We use only an originally supported mode of MTRR: setting the default cache mode to WP.
+    unsafe {
```

Comment:
Wrap this in a separate struct in the same way like these: https://docs.rs/x86_64/latest/x86_64/registers/model_specific/index.htm

The code would roughly look like this:
```
struct IA32MTRRdefType;

impl IA32MTRRdefType {
  const MSR: Msr = Msr(0xabcdef);
 
  pub fn write(enable: bool, fixed_range: bool, default_type: MemoryType) {
   // ...
  }
}
```